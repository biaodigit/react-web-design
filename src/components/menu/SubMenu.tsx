import React, { useContext, useState, FunctionComponentElement } from 'react';
import classNames from 'classnames';
import Transition from '../transition';
import Icon from '../icon/icon';
import { MenuContext } from './menu';
import { MenuItemProps } from './menuItem';

export interface SubMenuProps {
  index?: string;
  title: string;
  className?: string;
}

export const SubMenu: React.FC<SubMenuProps> = (props) => {
  const { index, title, className, children } = props;
  const context = useContext(MenuContext);
  const defaultOpenSubMenus = context.defaultOpenSubMenus as Array<string>;
  const isOpened =
    index && Object.is(context.mode, 'vertical')
      ? defaultOpenSubMenus.includes(index)
      : false;
  const [menuOpen, setOpen] = useState(isOpened);
  const classes = classNames('menu-item submenu-item', className, {
    'is-active': Object.is(context.index, index),
    'is-opened': menuOpen,
    'is-vertical': Object.is(context.mode, 'vertical'),
  });
  let timer: any;

  const handleClick = (e: React.MouseEvent) => {
    e.preventDefault();
    setOpen(!menuOpen);
  };

  const handleMouse = (e: React.MouseEvent, toggle: boolean) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      setOpen(toggle);
    }, 300);
  };

  const renderChildren = () => {
    const subMenuClasses = classNames('submenu', {
      'menu-opened': menuOpen,
    });
    const childrenComponent = React.Children.map(children, (child, i) => {
      const childElement = child as FunctionComponentElement<MenuItemProps>;
      const { displayName } = childElement.type;
      if (displayName === 'MenuItem') {
        return React.cloneElement(childElement, {
          index: `${index}-${i}`,
        });
      } else {
        console.error(
          'Warning:SubMenu has child which is not a MenuItem Component'
        );
      }
    });
    return (
      <Transition in={menuOpen} timeout={300} animation="zoom-in-top">
        <ul className={subMenuClasses}>{childrenComponent}</ul>
      </Transition>
    );
  };

  const clickEvents = Object.is(context.mode, 'vertical')
    ? { onClick: handleClick }
    : {};

  const hoverEvents = !Object.is(context.mode, 'vertical')
    ? {
        onMouseEnter: (e: React.MouseEvent) => handleMouse(e, true),
        onMouseLeave: (e: React.MouseEvent) => handleMouse(e, false),
      }
    : {};
  return (
    <li key={index} className={classes} {...hoverEvents}>
      <div className="submenu-title" {...clickEvents}>
        {title}
        <Icon icon="angle-down" className="arrow-icon" />
      </div>
      {renderChildren()}
    </li>
  );
};

SubMenu.displayName = 'SubMenu';

export default SubMenu;
