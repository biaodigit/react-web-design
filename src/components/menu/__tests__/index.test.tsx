import React from 'react';
import {
  render,
  RenderResult,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import Menu, { MenuProps } from '../Menu';
import MenuItem, { MenuItemProps } from '../MenuItem';
import SubMenu, { SubMenuProps } from '../SubMenu';

jest.mock('../../icon/Icon', () => {
  return () => {
    return <i className="fa" />;
  };
});
jest.mock('react-transition-group', () => {
  return {
    CSSTransition: (props: any) => {
      return props.children;
    },
  };
});

const testProps: MenuProps = {
  defaultIndex: '0',
  onSelect: jest.fn(),
  className: 'test',
};

const testVerticalProps: MenuProps = {
  defaultIndex: '0',
  mode: 'vertical',
  defaultOpenSubMenus: ['4'],
};

let wrapper: RenderResult,
  ver_wrapper: RenderResult,
  menuElement: HTMLElement,
  activeElement: HTMLElement,
  disabledElement: HTMLElement;

const generateMenu = (props: MenuProps) => (
  <Menu {...props}>
    <MenuItem>active</MenuItem>
    <MenuItem disabled>disable</MenuItem>
    <MenuItem>default</MenuItem>
    <SubMenu title="dropdown">
      <MenuItem>dropdown1</MenuItem>
    </SubMenu>
    <SubMenu title="opened">
      <MenuItem>opened2</MenuItem>
    </SubMenu>
  </Menu>
);

const createStyleFile = () => {
  const cssFile: string = `
     .submenu {
       display:none
     }
     .submenu.menu-opened {
       display:block
     }
  `;
  const style = document.createElement('style');
  style.innerHTML = cssFile;
  return style;
};

describe('test Menu Component', () => {
  beforeEach(() => {
    wrapper = render(generateMenu(testProps));
    wrapper.container.appendChild(createStyleFile());
    menuElement = wrapper.getByTestId('test-menu');
    activeElement = wrapper.getByText('active');
    disabledElement = wrapper.getByText('disable');
  });
  it('render snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('render default menu', () => {
    expect(menuElement).toBeInTheDocument();
    expect(menuElement).toHaveClass('menu test');
    expect(menuElement.querySelectorAll(':scope > li').length).toEqual(5);
    expect(activeElement).toHaveClass('menu-item is-active');
    expect(disabledElement).toHaveClass('menu-item is-disabled');
  });

  it('click items should change active and call the right callback', () => {
    const thirdItem = wrapper.getByText('default');
    fireEvent.click(thirdItem);
    expect(thirdItem).toHaveClass('is-active');
    expect(activeElement).not.toHaveClass('is-active');
    expect(testProps.onSelect).toHaveBeenCalledWith('2');
    fireEvent.click(disabledElement);
    expect(disabledElement).not.toHaveClass('is-active');
    expect(testProps.onSelect).not.toHaveBeenCalledWith('1');
  });

  it('should show dropdown items when hover on subMenu', async () => {
    expect(wrapper.queryByText('dropdown1')).not.toBeVisible();
    const dropdownElement = wrapper.getByText('dropdown');
    fireEvent.mouseEnter(dropdownElement);
    await waitFor(() => {
      expect(wrapper.getByText('dropdown1')).toBeVisible();
    });
    fireEvent.click(wrapper.getByText('dropdown1'));
    expect(testProps.onSelect).toHaveBeenCalledWith('3-0');
    fireEvent.mouseLeave(dropdownElement);
    await waitFor(() => {
      expect(wrapper.getByText('dropdown1')).not.toBeVisible();
    });
  });
});

describe('test Menu and MenuItem component in vertical mode', () => {
  beforeEach(() => {
    ver_wrapper = render(generateMenu(testVerticalProps));
    ver_wrapper.container.append(createStyleFile());
    menuElement = ver_wrapper.getByTestId('test-menu');
  });

  it('should render vertical mode when mode is set to vertical', () => {
    expect(menuElement).toHaveClass('menu-vertical');
  });

  it('should show dropdown items when click on subMenu for vertical mode', () => {
    const dropItem = ver_wrapper.queryByText('dropdown1');
    expect(dropItem).not.toBeVisible();
    const dropdownElement = ver_wrapper.getByText('dropdown');
    fireEvent.click(dropdownElement);
    expect(dropItem).toBeVisible();
  });

  it('should show subMenu dropdown when defaultOpenSubMenus contains SubMenu index', () => {
    expect(ver_wrapper.getByText('opened2')).toBeVisible();
  });
});
