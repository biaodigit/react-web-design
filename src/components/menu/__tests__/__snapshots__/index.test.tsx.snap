// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Menu Component render snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="web-menu test web-menu-horizontal"
        data-testid="test-menu"
      >
        <li
          class="web-menu-item is-active"
        >
          active
        </li>
        <li
          class="web-menu-item is-disabled"
        >
          disable
        </li>
        <li
          class="web-menu-item"
        >
          default
        </li>
        <li
          class="web-menu-item web-submenu-item"
        >
          <div
            class="web-submenu-title"
          >
            dropdown
            <i
              class="fa"
            />
          </div>
          <ul
            class="web-submenu"
          >
            <li
              class="web-menu-item"
            >
              dropdown1
            </li>
          </ul>
        </li>
        <li
          class="web-menu-item web-submenu-item"
        >
          <div
            class="web-submenu-title"
          >
            opened
            <i
              class="fa"
            />
          </div>
          <ul
            class="web-submenu"
          >
            <li
              class="web-menu-item"
            >
              opened2
            </li>
          </ul>
        </li>
      </ul>
      <style>
        
     .web-submenu {
       display:none
     }
     .web-submenu.web-menu-opened {
       display:block
     }
  
      </style>
    </div>
  </body>,
  "container": <div>
    <ul
      class="web-menu test web-menu-horizontal"
      data-testid="test-menu"
    >
      <li
        class="web-menu-item is-active"
      >
        active
      </li>
      <li
        class="web-menu-item is-disabled"
      >
        disable
      </li>
      <li
        class="web-menu-item"
      >
        default
      </li>
      <li
        class="web-menu-item web-submenu-item"
      >
        <div
          class="web-submenu-title"
        >
          dropdown
          <i
            class="fa"
          />
        </div>
        <ul
          class="web-submenu"
        >
          <li
            class="web-menu-item"
          >
            dropdown1
          </li>
        </ul>
      </li>
      <li
        class="web-menu-item web-submenu-item"
      >
        <div
          class="web-submenu-title"
        >
          opened
          <i
            class="fa"
          />
        </div>
        <ul
          class="web-submenu"
        >
          <li
            class="web-menu-item"
          >
            opened2
          </li>
        </ul>
      </li>
    </ul>
    <style>
      
     .web-submenu {
       display:none
     }
     .web-submenu.web-menu-opened {
       display:block
     }
  
    </style>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
